### ISSUE TITLE
tar completion fails with failglob

### ISSUE THREAD

**Issue by jwodder** [scop/bash-completion#152]
* Ubuntu 16.04.3 LTS
* GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)
* bash-completion version 2.7
* GNU tar 1.28

With failglob on:

```
$ ls -A
something.tar.gz
$ tar zxf <TAB>bash: no match: --occurrence[=NUMBER]
$
```

Completion works as expected when failglob is turned off.

For completeness, my `set -o` settings:

```
allexport       off
braceexpand     on
emacs           on
errexit         off
errtrace        off
functrace       off
hashall         on
histexpand      on
history         on
ignoreeof       on
interactive-comments    on
keyword         off
monitor         on
noclobber       off
noexec          off
noglob          off
nolog           off
notify          off
nounset         off
onecmd          off
physical        off
pipefail        on
posix           off
privileged      off
verbose         off
vi              off
xtrace          off
```

and my `shopt` settings:

```
autocd          off
cdable_vars     off
cdspell         off
checkhash       off
checkjobs       off
checkwinsize    on
cmdhist         on
compat31        off
compat32        off
compat40        off
compat41        off
compat42        off
complete_fullquote      on
direxpand       off
dirspell        off
dotglob         off
execfail        off
expand_aliases  on
extdebug        off
extglob         on
extquote        on
failglob        on
force_fignore   on
globstar        on
globasciiranges off
gnu_errfmt      off
histappend      off
histreedit      off
histverify      off
hostcomplete    off
huponexit       off
interactive_comments    on
lastpipe        off
lithist         off
login_shell     off
mailwarn        off
no_empty_cmd_completion on
nocaseglob      off
nocasematch     off
nullglob        off
progcomp        on
promptvars      on
restricted_shell        off
shift_verbose   off
sourcepath      on
xpg_echo        off
```

**Comment by dahart** [comment]
I get the same behavior in Ubuntu 18.04. Is the default _gtar completer expecting a different version of tar or something? My /bin/tar is gnu tar. Even 'complete -G * tar' is better than no completion.

I poked around and found something about tar -o <TAB> working, and that does work for me, but the -o option doesn't make sense for extracting a tar, and extracting a tar is my most common use case.

**Comment by rosyth** [comment]
It appears that this is all part of the hyperbloated 'completion'  setup we now get with linux.  You can try and fix something in the >700 lines of shell code devoted to 'tar' completion (look in '/usr/share/bash-completion/completions/tar').
My solution (not really a fix) is to defeat the whole 'tar' completion code,  by inserting the following at the top of the file...
```
function _tar { : ; }  # ie do nothing
complete -F _tar        -o dirnames tar
return
```
I can't say what side effect this may have.  Not been aware of any.
I posted this fix on,
https://forums.linuxmint.com/viewtopic.php?p=1952970#p1952970
But nobody commented so far either.

**Comment by diazona** [comment]
I have the same problem on Gentoo. I traced it to the function [`__gtar_parse_help_line()`](https://github.com/scop/bash-completion/blob/master/completions/tar#L80-L97). The function includes a loop that starts out like this:
```
    for i in $1; do
```
Since `$1` is unquoted, bash will perform various stages of expansion, including pathname expansion, on the value.

The function `__gtar_parse_help_line()` gets run for each line in the output of `tar --help`, and there are several lines which include patterns that bash recognizes as pathname expansions. The first of these is `--occurrence[=NUMBER]`. Normally bash would try to find a file named `--occurrence=`, `--occurrenceN`, or so on, and would just ignore it when it can't find any such file, but when `failglob` is set, the absence of a matching file leads to a failure.

I can think of two possible ways to fix this off the top of my head. One is to unset `failglob` while the loop is running, i.e.
```
    local i restore_failglob
    restore_failglob="$(shopt -p failglob)"
    shopt -u failglob
    for i in $1; do
        ...
    done
    eval "$restore_failglob"
```
The other option is to use a different method of performing word splitting, like `read`:
```
    local i arg="$1"
    while [[ -n "$arg" ]]; do
        read i arg <<<"$arg"
        ...
    done
```
Neither of these seem _particularly_ clean, but I suppose that's how it is with shell programming. Maybe there's a better method out there though.

**Comment by scop** [comment]
Excellent, addressed in 0fbb56617fd683cf31dce3fb5f663efc0979c4d2
